### PR TITLE
fix: introduce toolbar component

### LINKED ISSUE #840: New Component - Toolbar
Create the new Toolbar component from the Fiori 3 specs.  The toolbar will be used by several components including Table, List, Panel, Form, etc.  (13) (Time - 6d )

### PR DESCRIPTION
**Pull request by rengare**
## Related Issue
Closes #840 

## Description
PR brings fiori3 toolbar and title toolbar as new component

## Screenshots
image

image

### CODE REVIEW

File: docs/pages/components/toolbar.md [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+## Toolbar types
+
+### Solid
+
+{% capture toolbar %}
+
+  <div class="fd-toolbar fd-toolbar--active fd-toolbar--solid">
+    <span>Title</span>
+    <button class="fd-button fd-button--compact">Button</button>
```
Comment (mikerodonnell89) [inline docs/pages/components/toolbar.md:23]: These are transparent type buttons in the specs, I think we should match that design

File: docs/pages/components/button.md [hunk 2]
```diff
@@ -7,9 +7,11 @@ toc: false
 permalink: components/button.html
 folder: components
 ---
-{: .docs-intro}
+
```
Comment (droshev) [inline docs/pages/components/button.md:10]: this can be done in a separate PR

Comment (rengare) [reply docs/pages/components/button.md]: I removed change made to button.md

File: docs/pages/components/toolbar.md (hunk 1 above)
Comment (rengare) [reply docs/pages/components/toolbar.md]: Right, fixed ;)

File: src/toolbar.scss [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+$block: #{$fd-namespace}-toolbar;
+$toolbar-height-compact: 2rem;
+$toolbar-height-cozy: 2.75rem;
+
+$toolbar-separator-height-compact: 1.5rem;
+$toolbar-separator-height-cozy: 2rem;
+
+$toolbar-spacing: 0.5rem;
+$toolbar-padding: 0 0.5rem;
+$title-toolbar-padding: 0 0.1rem 0 0.5rem;
```
[the remark this answers is not in the document]
Comment (rengare) [reply src/toolbar.scss]: fixed, thanks :)

**Comment by prsdthkr** [comment]
Noticed that the entire toolbar gets the active state background on click. Not sure if this is by design.
toolbar

File: src/toolbar.scss [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+  @include toolbarSpacing();
+
+  &--title {
+    @include titleToolbar();
+  }
+
+  &--standard {
+    @include toolbarStandard();
+  }
+
```
Comment (InnaAtanasova) [inline src/toolbar.scss:123]: I might be completely wrong but how I read the specs I think the standard toolbar is `design` not a type. There should not be a `--standard` modifier class. All toolbars should have a border-bottom and if you want a toolbar without a border you need to apply the `--clear` modifier class. But again, I might be wrong.

File: src/toolbar.scss [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+  &__spacer {
+    height: 100%;
+
+    &--auto {
+      flex: 1 auto;
+    }
+
+    &--fixed {
+      display: inline-block;
+    }
```
Comment (InnaAtanasova) [inline src/toolbar.scss:160]: One of these should be the default one and the other should overwrite the behaviour. The less modifier classes the better :) I would put the auto behaviour as default and fixed as a modifier class (like it's now)

File: src/toolbar.scss [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+$block: #{$fd-namespace}-toolbar;
+$toolbar-height-compact: 2rem;
+$toolbar-height-cozy: 2.75rem;
+
+$toolbar-separator-height-compact: 1.5rem;
+$toolbar-separator-height-cozy: 2rem;
+
+$toolbar-spacing: 0.5rem;
+$toolbar-padding: 0 0.5rem;
+$title-toolbar-padding: 0 0.5rem 0 0.1rem;
```
Comment (InnaAtanasova) [inline src/toolbar.scss:13]: I believe the last value is 1rem

File: docs/pages/components/toolbar.md [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+{% endcapture %}
+{% include display-component.html component=toolbar %}
+
+### Title
+
+{% capture toolbar %}
+
+  <div class="fd-toolbar fd-toolbar--solid fd-toolbar--title fd-toolbar-active">
+    <span>Title</span>
```
Comment (InnaAtanasova) [inline docs/pages/components/toolbar.md:90]: Should be h4 element

File: src/toolbar.scss [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+@mixin activeStyles() {
+  &.#{$block}--active {
+    @content;
+  }
+}
+
+@mixin toolbarAuto() {
+  @include activeStyles() {
+    @extend %toolbarStandardColors;
```
Comment (InnaAtanasova) [inline src/toolbar.scss:34]: I believe the background color of Auto should be `inherit` since it's based on the parent

File: src/toolbar.scss [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+$title-toolbar-height: 2.75rem;
+
+%toolbarStandardColors {
+  @include fd-hover() {
+    background-color: var(--sapList_Hover_Background);
+  }
+
+  @include fd-active() {
+    background-color: var(--sapActiveColor);
+  }
```
Comment (InnaAtanasova) [inline src/toolbar.scss:23]: if you put these values in `fd-toolbar` (You don't need --standard modifier) the rest will inherit them so you won't need to include them in each toolbar type and you can use the modifier classes to overwrite these props. We always have a default behaviour (usually it's the standard type) and if we need to overwrite the default behaviour we rely on the modifier classes

Comment (rengare) [reply src/toolbar.scss:23]: fixed

File: src/toolbar.scss (hunk 8 above)
Comment (rengare) [reply src/toolbar.scss]: fixed

File: docs/pages/components/toolbar.md (hunk 7 above)
Comment (rengare) [reply docs/pages/components/toolbar.md]: fixed

File: src/toolbar.scss (hunk 6 above)
Comment (rengare) [reply src/toolbar.scss]: fixed

File: src/toolbar.scss (hunk 5 above)
Comment (rengare) [reply src/toolbar.scss]: fixed

File: src/toolbar.scss (hunk 4 above)
Comment (rengare) [reply src/toolbar.scss:114]: fixed

File: src/toolbar.scss [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+  &.#{$block}--active {
+    @content;
+  }
+}
+
+@mixin toolbarAuto() {
+  @include activeStyles() {
+    @extend %toolbarStandardColors;
+  }
+}
```
[the remark this answers is not in the document]
Comment (rengare) [reply src/toolbar.scss]: fixed, but I left this mixing just in case its gonna change in future

File: src/fundamental-styles.scss [hunk 11]
```diff
@@ -64,6 +64,7 @@
 @import "./tile";
 @import "./time";
 @import "./token";
+@import "./toolbar";
```
[the remark this answers is not in the document]
Comment (rengare) [reply src/fundamental-styles.scss]: fixed

File: docs/pages/components/toolbar.md [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+    <span class="fd-toolbar__spacer fd-toolbar__spacer--auto"></span>
+    <button>opt1</button>
```
[the remark this answers is not in the document]
Comment (rengare) [reply docs/pages/components/toolbar.md]: fixed

**Comment by rengare** [comment]
>     * You don't have to show LTR in the documentation, only RTL
> 
>     * The documentation is missing Cosy example
> 
>     * Is Overflow part of the scope of this issue?

Yep, I'll add this

**Comment by rengare** [comment]
> Noticed that the entire toolbar gets the active state background on click. Not sure if this is by design.
> 
> toolbar
> 
> Similarly, the `fd-toolbar--info` toolbar gets a darker background when clicked giving the impression that entire toolbar is clickable.

I have to talk with the team because this was my understanding of specs

File: src/toolbar.scss [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+@mixin toolbarSolid() {
+  background-color: var(--sapBackgroundColor);
+}
+
+@mixin toolbarTransparent() {
+  background-color: var(--sapToolbar_Background);
+}
+
+@mixin toolbarInfo() {
+  // no contrast function color: contrast(@sapUiInfobarBackground, @ sapUiBaseText , @ sapUiContentContrastTextColor , @sapUiContentContrastTextThreshold);
```
[the remark this answers is not in the document]
Comment (rengare) [reply src/toolbar.scss:45]: #1043

File: src/toolbar.scss [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+      flex: none;
+      display: inline-block;
+    }
+  }
+
+  &__separator {
+    width: 0.0625rem;
+    height: $toolbar-separator-height-compact;
+    // missing color
+    // background-color: var(@sapUiToolbarSeparatorColor);
```
[the remark this answers is not in the document]
Comment (rengare) [reply src/toolbar.scss:156]: #1044

File: stories/toolbar/toolbar.html [hunk 15]
```diff
@@ -0,0 +1,120 @@
+<script>
```
[the remark this answers is not in the document]
Comment (rengare) [reply stories/toolbar/toolbar.html:1]: I think it's doable but I'll do this in another PR

File: docs/pages/components/toolbar.md [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+    <button class="fd-button fd-button--compact fd-button--transparent sap-icon--survey"></button>
+    <button class="fd-button fd-button--compact fd-button--transparent sap-icon--pool"></button>
+    <button class="fd-button fd-button--compact fd-button--transparent sap-icon--pie-chart"></button>
+  </div>
+
+{% endcapture %}
+{% include display-component.html component=toolbar %}
+
+### Auto
+
```
[the remark this answers is not in the document]
Comment (rengare) [reply docs/pages/components/toolbar.md:39]: done

File: docs/pages/components/toolbar.md [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+    <button class="fd-button fd-button--compact fd-button--transparent sap-icon--survey"></button>
+    <button class="fd-button fd-button--compact fd-button--transparent sap-icon--pool"></button>
+    <button class="fd-button fd-button--compact fd-button--transparent sap-icon--pie-chart"></button>
+  </div>
+
+{% endcapture %}
+{% include display-component.html component=toolbar %}
+
+### Info
+
```
[the remark this answers is not in the document]
Comment (rengare) [reply docs/pages/components/toolbar.md:81]: done

File: docs/_data/sidebars/left-navigation-sidebar.yml [hunk 18]
```diff
@@ -240,6 +240,10 @@ entries:
       url: /components/token.html
       output: web
 
+    - title: Toolbar
```
[the remark this answers is not in the document]
Comment (rengare) [reply docs/_data/sidebars/left-navigation-sidebar.yml:243]: done

File: src/toolbar.scss (hunk 14 above)
[the remark this answers is not in the document]
Comment (rengare) [reply src/toolbar.scss:156]: #1044

File: src/toolbar.scss (hunk 13 above)
[the remark this answers is not in the document]
Comment (rengare) [reply src/toolbar.scss:45]: #1043

File: src/toolbar.scss [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+
+@mixin toolbarTransparent() {
+  background-color: var(--sapToolbar_Background);
+}
+
+@mixin toolbarInfo() {
+  // no contrast function color: contrast(@sapUiInfobarBackground, @ sapUiBaseText , @ sapUiContentContrastTextColor , @sapUiContentContrastTextThreshold);
+  $active-color: white;
+
+  // no contrast function: background: darken(@sapUiListBackground, 10);
```
[the remark this answers is not in the document]
Comment (rengare) [reply src/toolbar.scss:48]: #1045

File: docs/pages/components/toolbar.md [hunk 20]
```diff
@@ -0,0 +1,248 @@
...
+    <button class="fd-button fd-button--compact fd-button--transparent sap-icon--survey"></button>
+    <button class="fd-button fd-button--compact fd-button--transparent sap-icon--pool"></button>
+    <button class="fd-button fd-button--compact fd-button--transparent sap-icon--pie-chart"></button>
+  </div>
+
+{% endcapture %}
+{% include display-component.html component=toolbar %}
+
+### Transparent
+
```
[the remark this answers is not in the document]
Comment (rengare) [reply docs/pages/components/toolbar.md:60]: done

File: docs/pages/components/toolbar.md [hunk 21]
```diff
@@ -0,0 +1,248 @@
...
+    <button class="fd-button fd-button--compact fd-button--transparent sap-icon--pool"></button>
+    <button class="fd-button fd-button--compact fd-button--transparent sap-icon--pie-chart"></button>
+  </div>
+
+{% endcapture %}
+{% include display-component.html component=toolbar %}
+
+## Overflow
+
+The only addition OverflowToolbar has is the ability to place inout elements in a popover when
```
[the remark this answers is not in the document]
Comment (rengare) [reply docs/pages/components/toolbar.md]: done

File: src/toolbar.scss [hunk 22]
```diff
@@ -0,0 +1,159 @@
...
+  @include activeStyles() {
+    background-color: var(--sapInfobar_Background);
+    color: $active-color;
+
+    @include fd-hover() {
+      background-color: var(--sapInfobar_Hover_Background);
+    }
+
+    @include fd-active() {
+      background-color: var(--sapInfobar_Active_Background);
```
[the remark this answers is not in the document]
Comment (rengare) [reply src/toolbar.scss:60]: Specs don't say anything about active colour tbh, only information what colour should be used is provided for info toolbar

File: stories/toolbar/toolbar.html (hunk 15 above)
[the remark this answers is not in the document]
Comment (droshev) [reply stories/toolbar/toolbar.html:1]: sounds good, then you would need to introduce njk tests until we have this in storybook

File: docs/_data/components.yml [hunk 23]
```diff
@@ -1,97 +1,100 @@
 ## Data for componnets list display page.
-  items:
-    - title: Action Bar
-      meta: action-bar
+items:
```
Comment (droshev) [inline docs/_data/components.yml:2]: those formatting changes should be done in a separate PR

File: docs/pages/components/index.html [hunk 24]
```diff
@@ -9,11 +9,15 @@
...
 
 <div class="docs-component-grid">
-    {% for component in site.data.components.items %}
-        <a href="{{component.meta}}.html" class="docs-component-grid__item">
-            <img src="{{site.baseurl}}/images/component-visual/{{component.meta}}.png" alt="{{component.title}}" class="docs-component-grid__img docs-component-grid__img__{{component.meta}}">
-            <br>
-            <h3 class="docs-component-grid--label">{{component.title}}</h3>
-        </a>
-    {% endfor %}
+  {% for component in site.data.components.items %}
```
Comment (droshev) [inline docs/pages/components/index.html:12]: same as above

**Comment by InnaAtanasova** [comment]
Hey @rengare I don't see fd-reset() anywhere in the scss code. Usually this can cause bleeding of styling.

**Comment by droshev** [comment]
> Hey @rengare I don't see fd-reset() anywhere in the scss code. Usually this can cause bleeding of styling.

That's why I wrote that we should have njk test

File: docs/_data/components.yml (hunk 23 above)
Comment (rengare) [reply docs/_data/components.yml:2]: that means that prettierrc file is not used

File: docs/pages/components/index.html (hunk 24 above)
Comment (rengare) [reply docs/pages/components/index.html:12]: that means that prettierrc file is not used

**Comment by rengare** [comment]
> Hey @rengare I don't see fd-reset() anywhere in the scss code. Usually this can cause bleeding of styling.

Fixed, thanks :)

**Comment by rengare** [comment]
> > Hey @rengare I don't see fd-reset() anywhere in the scss code. Usually this can cause bleeding of styling.
> 
> That's why I wrote that we should have njk test

Our storybook has ability to run self-contained tests

**Comment by stefanoScalzo** [comment]
You need to add hover and active background colors for the toolbar auto

**Comment by stefanoScalzo** [comment]
Remove arrow on popover because it is dropdown variant

**Comment by stefanoScalzo** [comment]
a11y issue Ensures the order of headings is semantically correct
.

**Comment by rengare** [comment]
> You need to add hover and active background colors for the toolbar auto and for transparent toolbar

Hover and active colors are there when fd-toolbar--active class is presented

**Comment by InnaAtanasova** [comment]
Screen Shot 2020-05-22 at 3 09 36 PM

Just one minor thing

File: docs/pages/components/toolbar.md [hunk 25]
```diff
@@ -0,0 +1,248 @@
...
+    <span class="fd-toolbar__spacer fd-toolbar__spacer--auto"></span>
+    <button class="fd-button fd-button--compact fd-button--transparent">opt1</button>
+    <button class="fd-button fd-button--compact fd-button--transparent">opt2</button>
+  </div>
+
+{% endcapture %}
+{% include display-component.html component=toolbar %}
+
+## Active toolbar
+
```
[the remark this answers is not in the document]
Comment (prsdthkr) [reply docs/pages/components/toolbar.md:110]: I think the Active toolbar should be implemented as an InfoBar as per the specs here: https://experience.sap.com/fiori-design-web/infobar/
 According to my understanding 
1.  InfoBar is only supposed to have text
2. InfoBar cannot have other controls like buttons
3. InfoBar can be of two types: Active and Inactive
image
4. For Active InfoBar, the entire bar is clickable with keyboard and mouse.
5. All text content is visible, in all states (hover, focus, active) for all types (Active, inactive) of InfoBar

File: package.json [hunk 26]
```diff
@@ -130,4 +133,4 @@
     "IE 11"
   ],
   "dependencies": {}
-}
+}
```
Comment (prsdthkr) [inline package.json:136]: nit: add end-of-file newline

Comment (rengare) [reply package.json:136]: About Active toolbar its not true see attachment
image

**Comment by rengare** [comment]
@prsdthkr About active toolbar, other types of toolbar also have an active state

image

**Comment by rengare** [comment]
> Screen Shot 2020-05-22 at 3 09 36 PM
> 
> Just one minor thing



> Screen Shot 2020-05-22 at 3 09 36 PM
> 
> Just one minor thing

fixed, thanks :)

File: docs/pages/components/toolbar.md (hunk 25 above)
[the remark this answers is not in the document]
Comment (droshev) [reply docs/pages/components/toolbar.md:110]: Can we check/confirm this with the designers?
